Use PSID format in binding table

dpino opened the pull request:
This PR changes the format of each entry in the binding table:

```
{ b4_v6, v4, { psid_info }, br_v6 }
```

Where psid:

```
{ PSID, psid_offset, psid_length }
```

It also changes the lookup code in the binding table by converting a port number to PSID. Some tests are failing, probably it's necessary to update some pcap packets port number. This PR implements #101 and outdates #157.

dpino committed:
Lookup in binding table using PSID format

kbara commented on src/program/snabb_lwaftr/tests/data/binding.table:
```diff
@@ -1,7 +1,8 @@
- {'127:24:35:46:57:68:79:128', '178.79.150.2', 7800, 7900, '1E:1:1:1:1:1:1:af'},
- {'127:14:25:36:47:58:69:128', '178.79.150.3', 4000, 5050, '1E:2:2:2:2:2:2:af'}
+ {'127:2:3:4:5:6:7:128', '178.79.150.233', {0, 0, 9}, '8:9:a:b:c:d:e:f'},
+ {'127:11:12:13:14:15:16:128', '178.79.150.233', {0, 9, 0}},
```
It's invalid to have a different {a, k} (aka {offset, length}) pair for a given IPv4 address, as per http://tools.ietf.org/html/rfc7597 . ```To guarantee non-overlapping port sets, the offset 'a' MUST be the same for every MAP CE sharing the same address.``` (and the same for 'k').

kbara replied:
Probably the code should explicitly check for this when loading a table.

kbara replied:
The root cause is indeed a .233 rule. {'127:11:12:13:14:15:16:128', '178.79.150.233', {0, 9, 0}}, excludes only the A=0 ports (0-127), and includes the rest - including 64001, the port in the test, which is indeed using the .233 IPv4 address.

dpino commented:
@kbara I updated the PR. I changed the binding table, so it conforms to the spec. I managed to reduce the failing test cases to two. The failing test cases are ICMP related. Any idea on what could it be?

dpino commented:
Fixed one failing test. ICMP request/reply uses ICMP id as port. Updated PR.

kbara commented:
It'd be handy if you'd leave this kind of thing as multiple commits until it was time to merge it. I need to look over all the unrelated older changes to figure out what you've done now.

kbara commented:
```
 assert(offset == entry.offset and length == entry.length, 
            "There are at least two entries with the same IPv4 address but "..
            "with different PSID offset and PSID length")
```
This should probably explain that this is a violation of RFC 7597, which imposes the constraint to prevent overlapping port ranges.

kbara commented:
Wording details aside, the latest change looks good. The last failing test is due to a genuine lwaftr bug, which I'm investigating.

dpino commented:
@kbara sorry, agree amending makes it harder to read. I will append next time.

kbara commented:
Thank you.

dpino committed:
Enable 'HOP limit exceeded' test

dpino committed:
Use RangeMap to lookup IPv4's PSID offset and PSID length values

aperezdc commented:
Other than the comments from @andywingo and the small nit I commented, LGTM.

dpino committed:
Applied review fixes

andywingo commented on src/apps/lwaftr/lwaftr.lua:
```diff
@@ -43,6 +41,28 @@ local function compute_binding_table_by_ipv4(binding_table)
+      local key, value = extract_key_and_value(entry)
+      builder:add(key, value)
+   end
+   return builder:build()
```
Here can we patch the return result:

```lua
local map = builder:build()
function map:lookup_a_and_k(ipv4)
   local offset = self:lookup(ipv4)
   return self:val_at(offset).a, self:val_at(offset).k   
end
function map:lookup_psid(ipv4, port)
   local a, k = self:lookup_a_and_k(ipv4)
   return port_to_psid(port, a, k)
end
```

and adapt users.

andywingo commented on src/program/snabb_lwaftr/tests/data/binding.table:
```diff
@@ -1,7 +1,13 @@
- {'127:24:35:46:57:68:79:128', '178.79.150.2', 7800, 7900, '1E:1:1:1:1:1:1:af'},
- {'127:14:25:36:47:58:69:128', '178.79.150.3', 4000, 5050, '1E:2:2:2:2:2:2:af'}
+ {'127:2:3:4:5:6:7:128', '178.79.150.233', {80, 0, 16}, '8:9:a:b:c:d:e:f'},
+ {'127:11:12:13:14:15:16:128', '178.79.150.233', {2300, 0, 16}},
```
here can we specify these as psid=X, shift=S, length=X, with shift defaulting to 16-length ?

The positional names are too hard to read, and I think naming them A and K is not so great either.  I argue that "shift" and "length" are better names -- to get the PSID from a port, you shift right by SHIFT, and then take the lowest LENGTH bits.  With LENGTH taking its default value, we get contiguous port ranges.

Probably internally we should work on SHIFT and LENGTH instead of A and K.  The change is irritating, and I know this must be irritating to you.  I apologize.  What do you think about the suggestion though?  We could of course transform SHIFT/LENGTH into A/K, but to me there's no value in that.  I always forget which one is which.  That could just be me though.

If you agree, the format should be then e.g.

```
{'127:11:12:13:14:15:16:128', '178.79.150.233', {psid=2700, shift=0}},
```

or whatever.

kbara replied:
I don't want to call them 'shift' and 'length'; they are all lengths to me. I agree that a/k are overly cryptic.
Did we decide on forcing a = 0 and contiguous port ranges or not? If we did, it doesn't need a name.
k could reasonably be replaced with psid_len, I think.

andywingo replied:
I think we don't need to force a=0, given that we already have to look up the.... the what?  You see, we need names :)  I'll call it the bitwidth then.  "PSID bitwidth" -- how many bits wide the PSID is.  Anyway given that we have to look it up in a table, we might as well look up the shift at the same time.

We could alternately assume that shift = 16-bitwidth.  Probably a better way to do things.

dpino commented:
Agree on named parameters.

To recap, 'a' is bits shifted (SHIFT) and 'k' is PSID bitwidth (LENGTH). The names make sense to me as 'a' is defined as PSID offset (shift) and 'k' is defined as PSID length in the spec.

What I'm not sure is to leave SHIFT defined in terms of LENTH as 16-LENGTH, because this assumes m=0 (continue port range). If m=0, the range of ports is 1. What makes non-contiguous ports happen is a SHIFT different than zero, right? So as @kbara mentioned if we assume SHIFT is always zero, we only need LENGTH and can infer m out of LENGTH. So, format:

```
{'127:11:12:13:14:15:16:128', '178.79.150.233', {psid=2700, bitwidth=11}}
```

kbara commented:
I agree we need names for all three fields; I just really dislike the above ones.
I propose a = offset, following the use of http://tools.ietf.org/html/rfc7597#section-7.1 ; it's not great.
k = psid_len, in my head.
j = port_select_bits , or some variant that sounds a bit better. Most other names I can think of involve a base 2 log, which isn't the right direction to go with naming....

dpino commented:
@kbara I think we need only two parameters as the third one can be inferred out of the other two.'j' should be 'm' as we are talking about bits in the other cases. Quoting the spec:

```
                              0                   1
                              0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5
                             +-----------+-----------+-------+
               Ports in      |     A     |    PSID   |   j   |
            the CE port set  |    > 0    |           |       |
                             +-----------+-----------+-------+
                             |  a bits   |  k bits   |m bits |
```

I'm OK with `offset` and `psid_len` as named parameters. Format:

```
{'127:11:12:13:14:15:16:128', '178.79.150.233', {psid=2700, offset=0, psid_len=11}}
```
@wingo WDYT? Does those two parameter names sound good to you?

kbara commented:
We need to store two, but to -name- all three, for when we're discussing them, and secondarily for variables in our code. thingy = 16 - otherthingy - thirdthingy would be awkward, even if two of those names are good and are what we're storing.

andywingo commented:
For posterity, we just hashed this out over jabber and decided on these names: 
```
a: reserved_ports_bit_count
k: psid_len
m: shift
```

Looking forward to a new version of this PR with the new names :)

dpino committed:
Rename PSID's a, k, m attributes

andywingo commented on src/apps/lwaftr/binding_table.lua:
```diff
@@ -32,33 +32,72 @@ end
-      local pv4 = lwutil.rd32(ipv4:pton(v[2]))
+      local b4_v6, pv4, psid_info, br_v6 = unpack(v)
+      local reserved_ports_bit_count = psid_info.reserved_ports_bit_count or 0
+      local psid_len = psid_info.psid_len or 0
```
I think instead of specifying reserved_ports_bit_count and psid_len, we should specify psid_len and shift.  WDYT?

andywingo commented on src/apps/lwaftr/lwaftr.lua:
```diff
@@ -43,6 +44,43 @@ local function compute_binding_table_by_ipv4(binding_table)
+
+local function compute_psid_info_map(binding_table)
+   local value_type = ffi.typeof([[
+      struct { uint16_t reserved_ports_bit_count; uint16_t psid_len; }
```
here we should specify psid_len and shift, IMO.  i agree with widening these to 16 bits each to make the total entry size 8 bytes (with the 4 byte ipv4), excellent.

andywingo commented on src/apps/lwaftr/lwaftr.lua:
```diff
@@ -43,6 +44,43 @@ local function compute_binding_table_by_ipv4(binding_table)
+      builder:add(key, value)
+   end
+   local map = builder:build()
+   function map:lookup_value(ipv4)
```
I kinda think we should not allow the value to escape like this.  The reason is, the pointer is a heap object, and I am not sure that scalar replacement will do the right thing always.  That's why in a previous patch I suggested returning a and k directly (but really we should return k and m, or rather psid_len and shift), so we pretty much know that the value allocation will be removed.  Maybe it is OK though.

andywingo commented on src/apps/lwaftr/lwaftr.lua:
```diff
@@ -43,6 +44,43 @@ local function compute_binding_table_by_ipv4(binding_table)
 
+local function port_to_psid(port, reserved_ports_bit_count, psid_len)
+   local shift = 16 - (reserved_ports_bit_count + psid_len)
+   return band(rshift(port, shift), (2^psid_len)-1)
```
Let's change to take shift as an argument.  Let's also add a case to "assert" that if reserved_ports_bit_count is greater than 0 then we assign a bogus PSID so that the lookup will fail.  Like this:

```lua
local function port_to_psid(port, psid_len, shift)
   local psid_mask = bit.lshift(1, psid_len)-1
   local psid = bit.band(bit.rshift(port, shift), psid_mask)
   if psid_len + shift < 16 then
      local reserved_ports_bit_count = 16 - psid_len - shift
      local first_allocated_port = bit.lshift(1,reserved_ports_bit_count)
      if port < first_allocated_port then psid = psid_mask + 1 end
   end
   return psid
end
```

Pretty sure the impact of the test will be small in the usual case that reserved_ports_bit_count is 0.

kbara replied:
That sounds fragile.

andywingo replied:
I assume you refer to the practice of returning an invalid PSID when the port is in the restricted range, though your comment is short so I have to guess :)  I don't think this is fragile in any significant way (you didn't specify the way either :) and I also think this has an advantage, to wit:

This lookup is on the hot path.  If the port is invalid (in the restricted range), we need to indicate an error somewhere.  This function doesn't have the context to handle the error, but the code that can handle the error can't detect the error on its own.  Fine.  So we have to return a sentinel of some kind.  We could return two values or return optionally nil.  However I think this is not so great for something on the hot path.  Better to return a uint32 and punt the error on down the line -- the error is effectively "packet not found in binding table".  We have to check anyway whether the packet is in the binding table, so we can just give it a PSID that no tunnel will have -- and we know how to do that here.

kbara replied:
I agree a two-return convention isn't the thing to do here. We just need to be careful to never do something like mask it to psid_len bits.

andywingo replied:
what is the action to take from this review comment? :)

kbara replied:
"Be careful". :D
More pragmatically, I'd be tempted to do something like define INVALID_PSID=0xFFFFFFFF or something, and use that.

andywingo replied:
I get what you're saying.  I really appreciate that you think about all these things, but I do want to nudge you (and all of us) towards never making a comment that would leave the patch submitter wondering what to do.  "be careful" is not actionable :)

Let's split the difference and use local INVALID_PSID = 0x10000.  It's out of the uint16 range and in the int32 range which luajit tends to like better.  OK!

andywingo commented:
Looking good, just a couple nits.  Thank you Diego!

kbara commented on src/apps/lwaftr/binding_table.lua:
```diff
@@ -32,33 +32,72 @@ end
+         local entry = inserted[pv4]
+         assert(reserved_ports_bit_count == entry.reserved_ports_bit_count and psid_len == entry.psid_len, 
+            "There are at least two entries with the same IPv4 address but "..
+            "with different psid_info reserved_ports_bit_count and psid_info psid_len")
```
As before, I don't think this message is informative enough; see my previous comment on specifying the RFC and saying why it's a problem.

kbara commented on src/apps/lwaftr/binding_table.lua:
```diff
@@ -32,33 +32,72 @@ end
+
+Binding table format:
+   - Entry: {b4_v6, v4, { psid_info }, br_v6}
+   - PSID Info: {reserved_ports_bit_count=bits, psid_len=bits, shift=bits}
```
We're not storing all three. Please have the comment reflect the two we are storing, and mention that the third is computed.

dpino committed:
Set PSID shift instead of reserved_ports_bit_count

* Return bogus PSID if port is within restricted port range.
* Change port overlapping error message.

kbara commented:
I'd like to see a rough measure of the performance before and after this change before approving merging. If we're enforcing a (= reserved_ports_bit_count )= 0, which this code is, we're still using contiguous ranges, so it's hard to justify if it's slower.

dpino commented:
@kbara lwaftr_olive.csv vs lwaftr_olive_psid.csv.

https://gist.github.com/dpino/6a32bca1cb447bbff90a

 Environment: Intel(R) Xeon(R) CPU E5-2620 v3 @ 2.40GHz, L3: 15360K.

andywingo commented:
Those performance numbers before your change don't match what I just measured for the nutmeg branch.  Have we regressed in Olive or could it be you had the wrong NUMA affinities?

andywingo commented:
Measuring on that E5-2620v3 system, with nutmeg I get results like this:

```
[wingo@snabb1:~/snabbswitch/src]$ sudo ./snabb-lwaftr bench \
>     ../tests/apps/lwaftr/data/binding.table \
>     ../tests/apps/lwaftr/data/icmp_on_fail.conf \
>     ../tests/apps/lwaftr/benchdata/ipv4-0550.pcap \
>     ../tests/apps/lwaftr/benchdata/ipv6-0550.pcap
statisticsv6: 3.489 MPPS, 16.468 Gbps.
statisticsv4: 3.489 MPPS, 14.236 Gbps.
statisticsv6: 3.553 MPPS, 16.771 Gbps.
statisticsv4: 3.553 MPPS, 14.497 Gbps.
statisticsv6: 3.555 MPPS, 16.781 Gbps.
statisticsv4: 3.555 MPPS, 14.506 Gbps.
statisticsv6: 3.564 MPPS, 16.824 Gbps.
statisticsv4: 3.564 MPPS, 14.543 Gbps.
...
```

With olive:
```
[wingo@snabb1:~/snabbswitch/src]$ sudo ./snabb-lwaftr bench     program/snabb_lwaftr/tests/data/binding.table     program/snabb_lwaftr/tests/data/icmp_on_fail.conf program/snabb_lwaftr/tests/benchdata/ipv4-0550.pcap program/snabb_lwaftr/tests/benchdata/ipv6-0550.pcap
statisticsv6: 3.643 MPPS, 17.196 Gbps.
statisticsv4: 3.643 MPPS, 14.865 Gbps.
statisticsv6: 3.690 MPPS, 17.419 Gbps.
statisticsv4: 3.690 MPPS, 15.057 Gbps.
statisticsv6: 3.691 MPPS, 17.423 Gbps.
statisticsv4: 3.691 MPPS, 15.061 Gbps.
statisticsv6: 3.691 MPPS, 17.421 Gbps.
statisticsv4: 3.691 MPPS, 15.059 Gbps.
...
```

With lookup_psid:
```
[wingo@snabb1:~/snabbswitch/src]$ sudo ./snabb-lwaftr bench     program/snabb_lwaftr/tests/data/binding.table     program/snabb_lwaftr/tests/data/icmp_on_fail.conf program/snabb_lwaftr/tests/benchdata/ipv4-0550.pcap program/snabb_lwaftr/tests/benchdata/ipv6-0550.pcap
statisticsv6: 3.305 MPPS, 15.597 Gbps.
statisticsv4: 3.305 MPPS, 13.483 Gbps.
statisticsv6: 3.353 MPPS, 15.827 Gbps.
statisticsv4: 3.353 MPPS, 13.681 Gbps.
statisticsv6: 3.357 MPPS, 15.844 Gbps.
statisticsv4: 3.357 MPPS, 13.696 Gbps.
statisticsv6: 3.357 MPPS, 15.844 Gbps.
statisticsv4: 3.357 MPPS, 13.696 Gbps.
...
```

So olive is a marginal speedup relative to nutmeg, and lookup_psid is a regression.  We should investigate why this is and see how to improve.  Since this patch is smallish, we should be able to figure out where the bottleneck is and fix it before we land it or just after landing it.

andywingo commented:
Of course the benchmark program should be measuring nanoseconds and cycles per packet, but that's another issue :)

dpino commented:
@andywingo: Thanks for the new benchmarks. My previous benchmark results didn't consider CPU affinity so that was the issue.

kbara commented:
What's the plan with this. Are we merging it and fixing performance after, or is the merge blocked on the performance regression? @andywingo @dpino

andywingo commented:
Oh lordie, I don't know how I got those numbers before.  Perhaps it's crucial to taskset even when not doing NUMA-related things?  I don't know.  Anyway testing again, here are my results:

Nutmeg:
```
[wingo@snabb1:~/snabbswitch/src]$ sudo taskset -c 6 ./snabb-lwaftr bench      ../tests/apps/lwaftr/data/binding.table      ../tests/apps/lwaftr/data/icmp_on_fail.conf      ../tests/apps/lwaftr/benchdata/ipv4-0550.pcap      ../tests/apps/lwaftr/benchdata/ipv6-0550.pcap
statisticsv6: 3.866 MPPS, 18.247 Gbps.
statisticsv4: 3.866 MPPS, 15.774 Gbps.
statisticsv6: 3.888 MPPS, 18.351 Gbps.
statisticsv4: 3.888 MPPS, 15.863 Gbps.
statisticsv6: 3.876 MPPS, 18.296 Gbps.
statisticsv4: 3.876 MPPS, 15.815 Gbps.
statisticsv6: 3.884 MPPS, 18.334 Gbps.
statisticsv4: 3.884 MPPS, 15.848 Gbps.
statisticsv6: 3.890 MPPS, 18.363 Gbps.
statisticsv4: 3.890 MPPS, 15.873 Gbps.
statisticsv6: 3.883 MPPS, 18.326 Gbps.
statisticsv4: 3.883 MPPS, 15.841 Gbps.
statisticsv6: 3.879 MPPS, 18.310 Gbps.
statisticsv4: 3.879 MPPS, 15.828 Gbps.
statisticsv6: 3.891 MPPS, 18.366 Gbps.
statisticsv4: 3.891 MPPS, 15.875 Gbps.
statisticsv6: 3.889 MPPS, 18.356 Gbps.
statisticsv4: 3.889 MPPS, 15.867 Gbps.
^C
```

Olive:
```
[wingo@snabb1:~/snabbswitch/src]$ sudo taskset -c 6 ./snabb-lwaftr bench     program/snabb_lwaftr/tests/data/binding.table     program/snabb_lwaftr/tests/data/icmp_on_fail.conf program/snabb_lwaftr/tests/benchdata/ipv4-0550.pcap program/snabb_lwaftr/tests/benchdata/ipv6-0550.pcap
statisticsv6: 3.886 MPPS, 18.342 Gbps.
statisticsv4: 3.886 MPPS, 15.856 Gbps.
statisticsv6: 3.971 MPPS, 18.742 Gbps.
statisticsv4: 3.971 MPPS, 16.201 Gbps.
statisticsv6: 3.968 MPPS, 18.731 Gbps.
statisticsv4: 3.968 MPPS, 16.191 Gbps.
statisticsv6: 3.978 MPPS, 18.774 Gbps.
statisticsv4: 3.978 MPPS, 16.228 Gbps.
statisticsv6: 3.968 MPPS, 18.731 Gbps.
statisticsv4: 3.968 MPPS, 16.191 Gbps.
statisticsv6: 3.947 MPPS, 18.632 Gbps.
statisticsv4: 3.947 MPPS, 16.105 Gbps.
statisticsv6: 3.977 MPPS, 18.773 Gbps.
statisticsv4: 3.977 MPPS, 16.228 Gbps.
statisticsv6: 3.969 MPPS, 18.735 Gbps.
statisticsv4: 3.969 MPPS, 16.194 Gbps.
statisticsv6: 3.979 MPPS, 18.780 Gbps.
statisticsv4: 3.979 MPPS, 16.233 Gbps.
statisticsv6: 3.959 MPPS, 18.687 Gbps.
statisticsv4: 3.959 MPPS, 16.153 Gbps.
^C
```

Olive with #179:
```
[wingo@snabb1:~/snabbswitch/src]$ sudo taskset -c 6 ./snabb-lwaftr bench     program/snabb_lwaftr/tests/data/binding.table     program/snabb_lwaftr/tests/data/icmp_on_fail.conf program/snabb_lwaftr/tests/benchdata/ipv4-0550.pcap program/snabb_lwaftr/tests/benchdata/ipv6-0550.pcap
statisticsv6: 3.988 MPPS, 18.825 Gbps.
statisticsv4: 3.989 MPPS, 16.273 Gbps.
statisticsv6: 4.027 MPPS, 19.008 Gbps.
statisticsv4: 4.027 MPPS, 16.431 Gbps.
statisticsv6: 4.021 MPPS, 18.978 Gbps.
statisticsv4: 4.021 MPPS, 16.404 Gbps.
statisticsv6: 4.030 MPPS, 19.020 Gbps.
statisticsv4: 4.030 MPPS, 16.441 Gbps.
statisticsv6: 4.026 MPPS, 19.004 Gbps.
statisticsv4: 4.026 MPPS, 16.428 Gbps.
statisticsv6: 4.022 MPPS, 18.984 Gbps.
statisticsv4: 4.022 MPPS, 16.410 Gbps.
statisticsv6: 3.913 MPPS, 18.470 Gbps.
statisticsv4: 3.913 MPPS, 15.966 Gbps.
statisticsv6: 3.935 MPPS, 18.572 Gbps.
statisticsv4: 3.935 MPPS, 16.053 Gbps.
statisticsv6: 4.022 MPPS, 18.984 Gbps.
statisticsv4: 4.022 MPPS, 16.410 Gbps.
statisticsv6: 4.038 MPPS, 19.058 Gbps.
statisticsv4: 4.038 MPPS, 16.474 Gbps.
statisticsv6: 4.025 MPPS, 18.997 Gbps.
statisticsv4: 4.025 MPPS, 16.421 Gbps.
^C
```

This PR, rebased on top of #179:
```
[wingo@snabb1:~/snabbswitch/src]$ sudo taskset -c 6 ./snabb-lwaftr bench     program/snabb_lwaftr/tests/data/binding.table     program/snabb_lwaftr/tests/data/icmp_on_fail.conf program/snabb_lwaftr/tests/benchdata/ipv4-0550.pcap program/snabb_lwaftr/tests/benchdata/ipv6-0550.pcap
statisticsv6: 3.778 MPPS, 17.831 Gbps.
statisticsv4: 3.778 MPPS, 15.414 Gbps.
statisticsv6: 3.829 MPPS, 18.075 Gbps.
statisticsv4: 3.829 MPPS, 15.624 Gbps.
statisticsv6: 3.825 MPPS, 18.056 Gbps.
statisticsv4: 3.825 MPPS, 15.608 Gbps.
statisticsv6: 3.838 MPPS, 18.113 Gbps.
statisticsv4: 3.838 MPPS, 15.657 Gbps.
statisticsv6: 3.834 MPPS, 18.096 Gbps.
statisticsv4: 3.834 MPPS, 15.642 Gbps.
statisticsv6: 3.835 MPPS, 18.103 Gbps.
statisticsv4: 3.835 MPPS, 15.649 Gbps.
statisticsv6: 3.827 MPPS, 18.063 Gbps.
statisticsv4: 3.827 MPPS, 15.613 Gbps.
statisticsv6: 3.831 MPPS, 18.084 Gbps.
statisticsv4: 3.831 MPPS, 15.632 Gbps.
^C
```

So, we have a regression.  Not an enormous one but we should understand where it comes from, and I don't understand it yet.

andywingo commented:
I have rebased the remaining patches that don't include #179 and #181 to https://github.com/Igalia/snabbswitch/tree/wip-psid.  I am pretty sure that the additional overhead is related to the different way the binding table is laid out, that we have to recurse into {psid=foo, psid_len=bar} to check the PSID.

At the same time, having worked with this code a bit, I think we should be configuring the system in a different way.  We should have a separate configuration table which specifies the psid_len and shift values per IPv4 address.  This information should not be stored in the binding table at all.  The binding table should store just the PSID and not the parameters.  I propose we review and merge #179 and #181 and then work on this one in another PR, one in Igalia's repo that we can collaborate on.  WDYT dpino?

dpino commented:
@wingo: I agree with the proposed changes. I have reviewed and merged #179. I cannot review #181 as it's my own code :) Once #181 is reviewed and merged I can create a new PR out of the wip-psid branch. 

Moving out 'psid_len' and 'shift' so there's only a single PSID value in the binding table seems like a good idea.

dpino commented:
Closing issue as commits were moved to #185.
